stratum/tools/stratum-replay/stratum_replay.cc

```diff
@@ -0,0 +1,228 @@
+  }
+  auto channel = ::grpc::CreateChannel(FLAGS_grpc_addr, channel_credentials);
+  auto stub = ::p4::v1::P4Runtime::NewStub(channel);
+
+  // Sends the arbitration update with given device id and election id.
+  ::p4::v1::StreamMessageRequest stream_req;
+  std::vector<std::string> election_ids =
+      absl::StrSplit(FLAGS_election_id, ",");
+  CHECK_RETURN_IF_FALSE(election_ids.size() == 2) << "Invalid election ID.";
+  uint64 election_id_high;
+  uint64 election_id_low;
+  CHECK_RETURN_IF_FALSE(absl::SimpleAtoi(election_ids[0], &election_id_high))
+      << "Unable to parse string " << election_ids[0] << " to uint64";
+  CHECK_RETURN_IF_FALSE(absl::SimpleAtoi(election_ids[1], &election_id_low))
+      << "Unable to parse string " << election_ids[1] << " to uint64";
+  absl::uint128 election_id =
+      absl::MakeUint128(election_id_high, election_id_low);
+
+  stream_req.mutable_arbitration()->set_device_id(FLAGS_device_id);
+  stream_req.mutable_arbitration()->mutable_election_id()->set_high(
+      absl::Uint128High64(election_id));
+  stream_req.mutable_arbitration()->mutable_election_id()->set_low(
+      absl::Uint128Low64(election_id));
+
+  ::grpc::ClientContext context;
+  std::unique_ptr<ClientStreamChannelReaderWriter> stream =
+      stub->StreamChannel(&context);
+  if (!stream->Write(stream_req)) {
+    RETURN_ERROR(ERR_INTERNAL)
+        << "Failed to send request '" << stream_req.ShortDebugString()
+        << "' to switch.";
+  }
+
+  // Push the given pipeline config
+  ::p4::v1::SetForwardingPipelineConfigRequest fwd_pipe_cfg_req;
+  ::p4::v1::SetForwardingPipelineConfigResponse fwd_pipe_cfg_resp;
+  fwd_pipe_cfg_req.set_device_id(FLAGS_device_id);
+  fwd_pipe_cfg_req.mutable_election_id()->set_high(
+      absl::Uint128High64(election_id));
+  fwd_pipe_cfg_req.mutable_election_id()->set_low(
+      absl::Uint128Low64(election_id));
+  fwd_pipe_cfg_req.set_action(
+      ::p4::v1::SetForwardingPipelineConfigRequest::VERIFY_AND_COMMIT);
+  RETURN_IF_ERROR(ReadProtoFromTextFile(
+      FLAGS_p4info, fwd_pipe_cfg_req.mutable_config()->mutable_p4info()));
+  RETURN_IF_ERROR(ReadFileToString(
+      FLAGS_pipeline_cfg,
+      fwd_pipe_cfg_req.mutable_config()->mutable_p4_device_config()));
+  ::grpc::Status status;
+  {
+    ::grpc::ClientContext context;
+    status = stub->SetForwardingPipelineConfig(&context, fwd_pipe_cfg_req,
+                                               &fwd_pipe_cfg_resp);
+    CHECK_RETURN_IF_FALSE(status.ok())
+        << "Faild to push forwarding pipeline config: "
+        << ::stratum::hal::P4RuntimeGrpcStatusToString(status);
+  }
+
+  // Parse the P4Runtime write log file and send write requests to the
+  // target device.
+  std::string p4WriteLogs;
+  RETURN_IF_ERROR(::stratum::ReadFileToString(argv[1], &p4WriteLogs));
+  std::vector<std::string> lines =
+      absl::StrSplit(p4WriteLogs, '\n', absl::SkipEmpty());
+  for (std::string line : lines) {
```

Comment:
```suggestion
  for (const std::string& line : lines) {
```

Is there a need to copy the line?